app/src/components/features/rules/RuleBuilder/Header/ActionButtons/CreateRuleButton/actions/index.js

```diff
@@ -3,10 +3,104 @@ import { actions } from "../../../../../../../../store";
 //UTILS
 import { isValidUrl } from "../../../../../../../../utils/FormattingHelper";
 import { CONSTANTS as GLOBAL_CONSTANTS } from "@requestly/requestly-core";
-// LODASH
-import { inRange } from "lodash";
+import _, { inRange } from "lodash";
 import { getFunctions, httpsCallable } from "firebase/functions";
 import { ResponseRuleResourceType } from "types/rules";
+import {
+  extractInnerTextAndAttributesFromHTMLString,
+  getHTMLNodeName,
+  validateStringForSpecificHTMLNode,
+} from "./insertScriptValidators";
+import { isFeatureCompatible } from "utils/CompatibilityUtils";
+import FEATURES from "config/constants/sub/features";
+
+/**
+ * Convert user inputs into rule schema
+ *
+ * @param {*} ruleData
+ * @returns {
+ *  success: boolean,
+ *  newRuleData?: object, // in case of successful validation and transformation
+ *
+ *  // in case of failure
+ *  error?: string,
+ *  message?: string,
+ *  pairId?: string, // id of the pair that caused the error
+ * }
+ */
+export const transformAndValidateRuleFields = async (ruleData) => {
+  switch (ruleData.ruleType) {
+    case GLOBAL_CONSTANTS.RULE_TYPES.SCRIPT: {
+      const isCompatibleWithAttributesForScriptRule = isFeatureCompatible(FEATURES.SCRIPT_RULE_HTML_BLOCK);
+
+      if (!isCompatibleWithAttributesForScriptRule) {
+        return ruleData;
+      }
+
+      /* RuleData is explicitly read-only (reference to value in store) */
+      const ruleDataCopy = _.cloneDeep(ruleData);
+      const newPairs = [];
+
+      for (const pair of ruleDataCopy.pairs) {
+        const parsedPair = pair;
+        const newScripts = [];
+
+        for (const script of pair.scripts) {
+          let newScript = script;
+
+          const htmlNodeName = getHTMLNodeName(script.type, script.codeType);
+
+          if (script.type === GLOBAL_CONSTANTS.SCRIPT_TYPES.CODE) {
+            const res = await validateStringForSpecificHTMLNode(script.value, htmlNodeName);
+            if (!res.isValid) {
+              return {
+                error: res.validationError,
+                success: false,
+                message: res.errorMessage,
+                pairId: pair.id,
+              };
+            }
+            const { attributes, innerText: code } = extractInnerTextAndAttributesFromHTMLString(
+              script.value,
+              htmlNodeName
+            );
+            newScript = {
+              ...script,
+              attributes,
+              value: code,
+            };
```

Comment:
moving these checks to a separate function seemed like overkill since they were completely independent and also had to manipulate the `newRuleData` object in the parent scope. 

But I hope the rewrite justifies this logic staying here

I have renamed the functions according to the suggestions